pyccel/parser/semantic.py

```diff
@@ -3970,90 +3973,129 @@ def _visit_For(self, expr):
             for_expr.end_annotation = expr.end_annotation
         return for_expr
 
-
     def _visit_FunctionalFor(self, expr):
-        old_index   = expr.index
-        new_index   = self.scope.get_new_name()
-        # Identity comparison is required to only find duplicates of the index caused
-        # by the construction of FunctionalFor.
-        expr.substitute(old_index, new_index, is_equivalent = lambda x,y: x is y)
+        """
+        Visit and transform a FunctionalFor AST node into an equivalent code block.
 
-        target  = expr.expr
-        index   = new_index
-        indices = [self.scope.get_expected_name(i) for i in expr.indices]
-        dims    = []
-        body    = expr.loops[1]
+        This method processes a `FunctionalFor` expression and transforms the loop structure
+        into a corresponding code block.
 
-        idx_subs = {}
-        #scope = self.create_new_loop_scope()
+        Parameters
+        ----------
+        expr : pyccel.ast.functionalexpr.FunctionalFor
+            The FunctionalFor AST node.
 
-        # The symbols created to represent unknown valued objects are temporary
+        Returns
+        -------
+        pyccel.ast.basic.CodeBlock
+            A code block containing the equivalent loops and necessary variable allocations for the given `FunctionalFor` expression.
+        """
+        target  = expr.expr
+        indices = []
+        dims = []
+        idx_subs = {}
         tmp_used_names = self.scope.all_used_symbols.copy()
         i = 0
-        while isinstance(body, For):
+        loops = list(expr.loops)
+
+        # Inner function to handle PythonNativeInt variables
+        def handle_int_loop_variable(var_name, var_scope):
+            indices.append(var_name)
+            var = self._create_variable(var_name, PythonNativeInt(), None, {}, insertion_scope=var_scope)
+            dvar = self._infer_type(var)
+            return var, dvar
+
+        # Inner function to handle iterable variables
+        def handle_iterable_variable(var_name, element, var_scope):
+            indices.append(var_name)
+            dvar = self._infer_type(element)
+            class_type = dvar.pop('class_type')
+            if class_type.rank > 0:
+                class_type = class_type.switch_rank(class_type.rank - 1)
+                dvar['shape'] = dvar['shape'][1:]
+            if class_type.rank == 0:
```

Comment:
```suggestion
            elif class_type.rank == 0:
```

Reply:
Actually no, the rank is reduced by one in the first condition. both conditions need to visited sometimes